go/vt/vtgate/vstream_manager.go

```diff
@@ -415,18 +415,52 @@ func (vs *vstream) alignStreams(ctx context.Context, event *binlogdatapb.VEvent,
 	}
 }
 
-func (vs *vstream) getCells() []string {
+// getCells determines the availability zones to select tablets from.
+// 2 scenarios:
+//
+// 1. No cells specified by the client via the gRPC request.
+// Tablets from the local cell of the VTGate AND the cell alias that this cell belongs to will be selected by default
+// Local cell will take precedence.
+//
+// 2. Cells are specified by the client via the gRPC request
+// These cels will take precendence over the default local cell and its alias
+// and only tablets belonging to the specified cells will be selected.
+// If the "local:" tag is passed in as an option in the list of optCells,
+// the local cell of the VTGate will take precedence over any other cell specified.
+func (vs *vstream) getCells(ctx context.Context) []string {
 	var cells []string
 	if vs.optCells != "" {
-		for _, cell := range strings.Split(strings.TrimSpace(vs.optCells), ",") {
+		for i, cell := range strings.Split(strings.TrimSpace(vs.optCells), ",") {
+			// if the local tag is passed in, we must give local cell priority
+			// during tablet selection. Append the VTGate's local cell to the list of cells
+			if i == 0 && cell == "local:" {
```

Comment:
The local cell related changes in the TabletPicker also affect vttablet<->vttablet vstreams, but there we do not have this local cell correction. How are we going to document that in the [Tablet Picker docs](https://vitess.io/docs/15.0/reference/vreplication/tablet_selection/)?

I'd rather us do the same for vreplication: https://github.com/vitessio/vitess/blob/main/go/vt/vttablet/tabletmanager/vreplication/controller.go#L126 and for vdiff: https://github.com/vitessio/vitess/blob/main/go/vt/vttablet/tabletmanager/vdiff/table_differ.go#LL252

As you can see though, this starts to get tricker as it's being done in multiple places — which could unintentionally drift over time — and we may add new locations in the future. So what we should do instead, IMO, is move this local cell correction code to the TabletPicker itself. That discovery component will be running in the vtgate and vttablet, and it knows what its local cell is. Then this correction code would also be covered by the new unit tests and we can confirm the potential "lost cell" issue noted above.

Reply:
> Wouldn't optCells e.g. be: local:zone1a,zone1b,zone1c? In that case, wouldn't the first element be local:zone1a rather than local:? 

No, since the client has no way of telling us that what the local cell actually is, we decided on tagging just "local" to indicate local cell preference. @rohit-nayak-ps suggested making this "local:" to keep consistent with the hint in the tablet picker. 

I get your point about moving this into the tabletpicker though. Earlier we had the flag, but now we don't so it makes more sense for this to live there. I can update.

Reply:
Ahh, OK. I missed that. Sorry. 

That syntax feels awkward to me and is not consistent with `in_order:<type>,...`. It's similar but different — `local:,<cell>,...` — which I don't like. That feels like one small part of a bad UX to me (non-intuitive). We could align both though and make it more clear (IMO) that it's a general preference and not related to the cell immediately following it, if we used this syntax (a variant of what I shared above could parse this correctly):
`local_preference:<cell1>,<cell2>,<cell3>,....`  

I think that keyword conveys the idea better and maintains the same kind of syntax as `in_order:<type1>,<type2>,...`. 

If you both prefer the current syntax, then I think we could make it more clear that this is a dynamic placeholder by using something like:
`{{local_cell}},cell1,cell2,...`.

What do you both think?

Then the next issue is that the TabletPicker itself doesn't have direct access to anything it can use to determine the local cell it's in. So that's something that the callers (vtgate and vttablet) — who do have that info easily accessible — would have to pass in.